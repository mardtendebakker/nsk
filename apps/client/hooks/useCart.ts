import { useEffect, useState } from 'react';
import cartStore, {
  ADD_MODULE_REQUEST,
  CLEAR_CART_REQUEST,
  Cart, EVENTS, Module, REMOVE_MODULE_REQUEST,
} from '../stores/cart';

const useCart = (): {
  state: Cart,
  removeModule:(module: Module)=> void,
  addModule:(module: Module)=> void,
  clearCart:()=> void,
  totalModulesCount: () => number
  totalAmount: () => number
} => {
  const [state, setState] = useState<Cart>(cartStore.cart);

  function cb() {
    setState(cartStore.cart);
  }

  useEffect(() => {
    EVENTS.forEach((event) => {
      cartStore.on(event, cb);
    });

    return () => {
      EVENTS.forEach((event) => {
        cartStore.removeListener(event, cb);
      });
    };
  }, []);

  return {
    state,
    removeModule: (module: Module) => {
      cartStore.emit(REMOVE_MODULE_REQUEST, module);
    },
    addModule: (module: Module) => {
      cartStore.emit(ADD_MODULE_REQUEST, module);
    },
    clearCart: () => {
      cartStore.emit(CLEAR_CART_REQUEST);
    },
    totalModulesCount: () => state.modules.length,
    totalAmount: () => {
      let total = 0;

      state.modules.forEach(({ price }) => {
        total += price;
      });

      return total;
    },
  };
};

export default useCart;
