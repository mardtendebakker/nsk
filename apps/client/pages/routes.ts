export const DASHBOARD = '/';
export const SIGN_IN = '/sign-in';
