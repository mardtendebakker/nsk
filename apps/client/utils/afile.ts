import { AFile } from './axios/models/aFile';

export function buildAFileLink(afile: AFile): string {
  return `https://${afile.unique_server_filename}.s3.amazonaws.com/${afile.discr}/${afile.original_client_filename}`;
}
