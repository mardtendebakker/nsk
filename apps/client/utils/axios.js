import axios from 'axios';

export { AxiosPromise, AxiosResponse, AxiosError } from 'axios';

const baseURL = process.env.NODE_ENV === 'development' ? 'http://localhost:3333' : 'production_url';

export const client = axios.create({
  baseURL,
  timeout: 120000,
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
});

export const { CancelToken } = axios;

export const SUPLIERS_PATH = '';

export default client;
