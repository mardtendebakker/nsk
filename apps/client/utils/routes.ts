export const DASHBOARD = '/';
export const SIGN_IN = '/sign-in';
export const CUSTOMERS_CONTACTS = '/customers/contacts';
export const CUSTOMERS_CONTACTS_NEW = '/customers/contacts/new';
export const CUSTOMERS_EMAILS_NEW = '/customers/emails/new';
export const CUSTOMERS_EMAILS = '/customers/emails';
export const SUPPLIERS = '/suppliers/:id';
export const ACCOUNT_VERIFICATION = '/account-verification';
export const SETTINGS = '/settings';
export const ORDERS_PURCHASES = '/orders/purchases';
export const ORDERS_SALES = '/orders/sales';
export const STOCKS_PRODUCTS = '/stock/products';
export const STOCKS_REPAIR_SERVICES = '/stock/repair-services';
