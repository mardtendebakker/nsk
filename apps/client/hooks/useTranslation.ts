import { useEffect, useState } from 'react';
import { trans, setConfig } from 'itranslator';
import { setDefaultOptions } from 'date-fns';
import {
  nl, enUS, de, fr, ar, faIR,
} from 'date-fns/locale';
import Config from 'itranslator/lib/type/config';
import { getDefaultLocale, setDefaultLocale } from '../utils/storage';
import EventEmitter from '../utils/eventEmitter';
import nlSource from '../public/translations/nl';
import enSource from '../public/translations/en';
import deSource from '../public/translations/de';
import frSource from '../public/translations/fr';
import arSource from '../public/translations/ar';
import faSource from '../public/translations/fa';

export const LOCALE_UPDATED = 'LOCALE_UPDATED';
export type Trans = (key: string, config?: Config | undefined) => string;

export const localeMapping = {
  nl: nlSource,
  en: enSource,
  de: deSource,
  ar: arSource,
  fr: frSource,
  fa: faSource,
};

export const dateFnsMapping = {
  nl,
  en: enUS,
  de,
  fr,
  ar,
  fa: faIR,
};

class LocaleEmitter extends EventEmitter {
  #locale = getDefaultLocale() || 'nl';

  constructor() {
    super();
    setDefaultOptions({ locale: dateFnsMapping[this.#locale] });
  }

  get locale() { return this.#locale; }

  emit(event: string, payload: string) {
    if (LOCALE_UPDATED === event) {
      this.#locale = payload;

      setConfig({ source: localeMapping[payload] });
      setDefaultOptions({ locale: dateFnsMapping[payload] });
      setDefaultLocale(payload);
    }

    return super.emit(event, payload);
  }
}

export const localeStore = new LocaleEmitter();

const useTranslation = (): { updateLocale: (locale: string)=> void, trans: Trans, locale: string } => {
  const [locale, setLocale] = useState<string>(localeStore.locale);

  useEffect(() => {
    localeStore.on(LOCALE_UPDATED, setLocale);
    return () => {
      localeStore.removeListener(LOCALE_UPDATED, setLocale);
    };
  }, [setLocale]);

  const updateLocale = (newLocale: string) => {
    localeStore.emit(LOCALE_UPDATED, newLocale);
    setLocale(newLocale);
  };

  return {
    updateLocale,
    trans,
    locale,
  };
};

export default useTranslation;
