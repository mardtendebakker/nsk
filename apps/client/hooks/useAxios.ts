import { useEffect, useRef, useState } from 'react';
import { useSnackbar } from 'notistack';
import { trans } from 'itranslator';
import axios, {
  AxiosResponse, AxiosPromise, CancelToken, AxiosError, CanceledError,
} from '../utils/axios';
import { hideProgress, showProgress } from '../components/topLinearProgress';

export const GET = 'get';
export const POST = 'post';
export const PATCH = 'patch';
export const DELETE = 'delete';
export const PUT = 'put';

type Method = 'get' | 'post' | 'patch' | 'delete' | 'put';
type ResponseType = 'json' | 'blob';

const useAxios = <Data>(
  method: Method,
  path: string,
  {
    withProgressBar = false,
    showErrorMessage = true,
    showSuccessMessage = false,
    customSuccessMessage,
    defaultParams = {},
    customStatusesMessages,
  }
  : {
    withProgressBar? : boolean,
    showErrorMessage? : boolean,
    showSuccessMessage? : boolean,
    customSuccessMessage?: (response: AxiosResponse) => string
    defaultParams?: object,
    customStatusesMessages?: { [key: number]: string }
  } = {},
) : {
    data?: Data,
    response?: AxiosResponse,
    error?: AxiosError,
    performing: boolean,
    call: Call,
  } => {
  const [response, setResponse] = useState<AxiosResponse>();
  const [error, setError] = useState<AxiosError>();
  const [performing, setPerforming] = useState<boolean>(false);
  const { enqueueSnackbar } = useSnackbar();
  const source = useRef(CancelToken.source());

  useEffect(() => source.current.cancel, []);

  async function call(
    explicitPath: string,
    params?: object,
    body?: object,
    responseType?: ResponseType,
    headers?: object,
  )
    : AxiosPromise<AxiosResponse> {
    if (method === POST || method === PATCH || method === PUT) {
      return axios[method](explicitPath, body, {
        cancelToken: source.current.token,
        params: { ...defaultParams, ...params },
        responseType,
        headers,
      });
    }

    return axios[method](explicitPath, {
      cancelToken: source.current.token, params: { ...defaultParams, ...params }, data: body, responseType,
    });
  }

  function handleSuccess(handledResponse: AxiosResponse) {
    enqueueSnackbar(
      customSuccessMessage ? customSuccessMessage(handledResponse) : trans(handledResponse.status.toString()),
      { variant: 'success' },
    );
  }

  function handleError(e: Error | AxiosError) {
    const errorResponse = (e instanceof AxiosError && e.response) ? e.response : { status: 500, data: { message: undefined } };
    const errorMessage = customStatusesMessages && customStatusesMessages[errorResponse.status];
    let messages: string[] = [];

    switch (true) {
      case typeof errorMessage == 'string':
        messages.push(errorMessage);
        break;
      case Array.isArray(errorResponse.data.message):
        messages = errorResponse.data.message;
        break;
      case typeof errorResponse.data.message == 'string' && !!errorResponse.data.message:
        messages.push(errorResponse.data.message);
        break;
      default:
        messages.push(trans(errorResponse.status.toString()));
    }

    messages.forEach((message) => {
      enqueueSnackbar(message, { variant: 'error' });
    });
  }

  return {
    data: response?.data,
    response,
    error,
    performing,
    /** @throws {Error} */
    call: async (
      {
        params, body, path: explicitPath, responseType, headers,
      }
      : { params?: object, body?: object, path?: string, responseType?: ResponseType, headers?: object }
      = {},
      cb?: (e: Error, axiosResponse?: AxiosResponse) => void,
    ): Promise<AxiosResponse | void> => {
      try {
        const finalPath = explicitPath || path;

        setPerforming(true);
        if (withProgressBar) {
          showProgress();
        }

        const resp = await call(finalPath, params, body, responseType, headers);
        setResponse(resp);

        if (showSuccessMessage || customSuccessMessage) {
          handleSuccess(resp);
        }

        if (cb) {
          cb(null, resp);
        }

        return resp;
      } catch (e) {
        if (e instanceof CanceledError) {
          throw new Error(explicitPath + path);
        }

        setError(e);
        if (showErrorMessage) {
          handleError(e);
        }

        if (cb) {
          cb(e);
        } else {
          throw e;
        }
      } finally {
        setPerforming(false);

        if (withProgressBar) {
          hideProgress();
        }
      }

      return undefined;
    },
  };
};

export type Call = (
  arg0?: { params?: object, body?: object, path?: string, responseType?: ResponseType, headers?: object },
  cb?: (e: Error, axiosResponse?: AxiosResponse) => void,
) => Promise<AxiosResponse | void>;

export default useAxios;
