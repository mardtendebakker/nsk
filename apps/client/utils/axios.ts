import axios from 'axios';

export { AxiosError } from 'axios';
export type { AxiosPromise, AxiosResponse } from 'axios';

const baseURL = process.env.NODE_ENV === 'development' ? 'http://localhost:3333/api' : process.env.NX_AXIOS_BASE_URL;

export const client = axios.create({
  baseURL,
  timeout: 120000,
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
});

export const { CancelToken } = axios;

export const CUSTOMERS_PATH = '/customers/:id';
export const SUPPLIERS_PATH = '/suppliers/:id';

export default client;

export interface Customer {
  id: number,
  name?: string,
  representative?: string,
  kvk_nr?: string,
  email?: string,
  phone?: string,
  phone2?: string,
  street?: string,
  street_extra?: string,
  city?: string,
  country?: string,
  state?: string,
  zip?: string,
  street2?: string,
  street_extra2?: string,
  city2?: string,
  country2?: string,
  state2?: string,
  zip2?: string,
  is_partner?: number,
}

export interface Supplier {
  id: number,
  name?: string,
  representative?: string,
  email?: string,
  phone?: string,
  phone2?: string,
  street?: string,
  street_extra?: string,
  city?: string,
  country?: string,
  state?: string,
  zip?: string,
  street2?: string,
  street_extra2?: string,
  city2?: string,
  country2?: string,
  state2?: string,
  zip2?: string,
  partner?: string,
}
