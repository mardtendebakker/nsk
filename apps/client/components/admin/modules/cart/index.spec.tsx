import {
  render, screen, fireEvent, waitFor,
} from '@testing-library/react';
import { useRouter } from 'next/router';
import Cart from '.';
import useCart from '../../../../hooks/useCart';
import useAxios from '../../../../hooks/useAxios';
import useTranslation from '../../../../hooks/useTranslation';
import { ADMIN_MODULES_PAYMENTS } from '../../../../utils/routes';

jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}));

jest.mock('../../../../hooks/useCart', () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock('../../../../hooks/useAxios', () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock('../../../../hooks/useTranslation', () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock('./moduleLine', () => ({
  __esModule: true,
  default: function MockModuleLine({ module }) {
    return <div data-testid={`module-${module.id}`}>{module.name}</div>;
  },
}));

describe('Cart Component', () => {
  const mockRouter = {
    push: jest.fn(),
  };

  const mockTranslations = {
    cart: 'Shopping Cart',
    total: 'Total',
    pay: 'Pay Now',
    emptyCartMessage: 'Your cart is empty',
  };

  const mockModules = [
    { id: 1, name: 'Module 1', price: 100 },
    { id: 2, name: 'Module 2', price: 200 },
  ];

  const mockAxiosCall = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();

    (useRouter as jest.Mock).mockReturnValue(mockRouter);

    (useTranslation as jest.Mock).mockReturnValue({
      trans: (key: string) => mockTranslations[key],
    });

    (useCart as jest.Mock).mockReturnValue({
      state: { modules: [] },
      totalAmount: () => 0,
    });

    (useAxios as jest.Mock).mockReturnValue({
      call: mockAxiosCall,
      performing: false,
    });
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<Cart />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders empty cart message when no modules present', () => {
    render(<Cart />);

    expect(screen.getByText(mockTranslations.emptyCartMessage)).toBeInTheDocument();
    expect(screen.queryByText(mockTranslations.total)).not.toBeInTheDocument();
  });

  it('renders cart with modules and total when modules are present', () => {
    (useCart as jest.Mock).mockReturnValue({
      state: { modules: mockModules },
      totalAmount: () => 300,
    });

    render(<Cart />);

    mockModules.forEach((module) => {
      expect(screen.getByTestId(`module-${module.id}`)).toBeInTheDocument();
    });

    expect(screen.getByText(mockTranslations.total)).toBeInTheDocument();
    expect(screen.getByText('€ 300,00')).toBeInTheDocument();
  });

  it('handles payment flow correctly', async () => {
    const totalAmount = 300;
    const paymentUrl = 'https://payment.example.com';
    (useCart as jest.Mock).mockReturnValue({
      state: { modules: mockModules },
      totalAmount: () => totalAmount,
    });

    mockAxiosCall.mockResolvedValue({ data: paymentUrl });

    render(<Cart />);

    const payButton = screen.getByText(mockTranslations.pay);
    fireEvent.click(payButton);

    expect(mockAxiosCall).toHaveBeenCalledWith({
      body: {
        moduleIds: mockModules.map((m) => m.id),
        redirectUrl: `${window.location.origin}${ADMIN_MODULES_PAYMENTS}?clearCart=1`,
      },
    });

    await waitFor(() => {
      expect(mockRouter.push).toHaveBeenCalledWith(paymentUrl);
    });
  });

  it('disables pay button while payment is processing', () => {
    (useCart as jest.Mock).mockReturnValue({
      state: { modules: mockModules },
      totalAmount: () => 300,
    });

    (useAxios as jest.Mock).mockReturnValue({
      call: mockAxiosCall,
      performing: true,
    });

    render(<Cart />);

    const payButton = screen.getByText(mockTranslations.pay);
    expect(payButton).toBeDisabled();
  });

  it('formats total price correctly', () => {
    (useCart as jest.Mock).mockReturnValue({
      state: { modules: mockModules },
      totalAmount: () => 1234.56,
    });

    render(<Cart />);

    expect(screen.getByText('€ 1.234,56')).toBeInTheDocument();
  });
});
