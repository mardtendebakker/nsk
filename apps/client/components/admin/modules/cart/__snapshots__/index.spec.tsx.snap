// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Component matches snapshot 1`] = `
<DocumentFragment>
  <h4
    class="MuiTypography-root MuiTypography-h4 css-4il6ew-MuiTypography-root"
  >
    Shopping Cart
  </h4>
  <p
    class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
    Your cart is empty
  </p>
</DocumentFragment>
`;
